### PR TITLE
GPTQ vector

### PR DESCRIPTION
**Pull request by sugunav14**
### What does this PR do?

Type of change: ? 

### Usage

```python
# Add a code snippet demonstrating how to use this
```

### Before your PR is "*Ready for review*"

Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md) and your commits are signed (`git commit -s -S`).

Make sure you read and follow the [Security Best Practices](https://github.com/NVIDIA/Model-Optimizer/blob/main/SECURITY.md#security-coding-practices-for-contributors) (e.g. avoiding hardcoded `trust_remote_code=True`, `torch.load(..., weights_only=False)`, `pickle`, etc.).

- Is this change backward compatible?: ✅ / ❌ / N/A 
- If you copied code from any other sources or added a new PIP dependency, did you follow guidance in `CONTRIBUTING.md`: ✅ / ❌ / N/A 
- Did you write any new necessary tests?: ✅ / ❌ / N/A 
- Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?: ✅ / ❌ / N/A 

## Summary by CodeRabbit

* **New Features**
  * Added backend-specific GPTQ helper registration to allow backend-tailored GPTQ behavior.

* **Bug Fixes**
  * Prevented KV-cache state from leaking across repeated per-layer forwards during calibration.

* **Tests**
  * Added GPU-focused tests validating GPTQ combined with vector quantization, including accuracy and end-to-end comparisons.

### CODE REVIEW

**Review by cjluo-nv — COMMENTED** [review]
The PR adds GPTQ support for vector LUT quantization with several bug fixes. While the core algorithm and tests look solid, there are several issues to address:

1. **Bare `assert` for runtime validation** in `_blockwise_vector_update` — these should be `RuntimeError`/`ValueError` since asserts can be optimized away with `-O`.
2. **Tests depend on undocumented `luts` module** — all 3 tests import from `luts` which is not a standard package and will fail without it. Tests need skip markers or the dependency needs to be documented.
3. **`getattr(mtq, args.qformat)` fallback** in `hf_ptq.py` allows arbitrary attribute access — should validate the returned value is actually a valid quant config dict.
4. **Inconsistent `dtype` → `torch_dtype` fix** — one instance of `dtype="auto"` on line ~641 of `example_utils.py` was not updated.
5. **`weight_quantizer.disable()` after GPTQ is a broad behavioral change** affecting all GPTQ paths — this needs to be explicitly called out as it may affect downstream code that expects the quantizer to still be enabled after GPTQ.

File: modelopt/torch/quantization/utils/calib_utils.py [hunk 1]
```diff
@@ -224,6 +244,58 @@ def _blockwise_update(self, block_size):
...
+        scales (``_psx_scales``) on the quantizer via the backend's
+        ``static_scales`` path.  The GPTQ loop then slices per-vector-group
+        scales from ``_psx_scales`` for each sub-vector quantization call.
+        """
+        import torch.nn.functional as F
+
+        quantizer = self.module.weight_quantizer
+        assert quantizer.backend_extra_args.get("static_scales", False), (
+            "GPTQ vector update requires static_scales=True in backend_extra_args."
+        )
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/utils/calib_utils.py]: Bug (per codebase guidelines): Bare `assert` statements are used for runtime validation here and throughout `_blockwise_vector_update`. These can be silently skipped with `python -O`. Replace with explicit exceptions:

```python
if not quantizer.backend_extra_args.get("static_scales", False):
    raise RuntimeError("GPTQ vector update requires static_scales=True in backend_extra_args.")
```

Same applies to lines 266-267 (`assert self.weight ...`) and 268 (`assert block_size % quant_block_size == 0`).

File: modelopt/torch/quantization/utils/calib_utils.py [hunk 2]
```diff
@@ -134,6 +135,22 @@ def free(self):
...
+        quantizer = self.module.weight_quantizer
+        original = quantizer._psx_scales
+        quantizer._psx_scales = scales
+        try:
+            yield
+        finally:
+            quantizer._psx_scales = original
+
     def update_weights(self, block_size, perc_damp):
         """Run GPTQ blockwise weight update on this module.
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/utils/calib_utils.py:138]: Nit: `is_vector_lut()` accesses `self.module.weight_quantizer` without guarding for its existence. While `hasattr` is used for `backend_extra_args`, if the module has no `weight_quantizer` attribute at all, this will raise `AttributeError`. Consider adding a guard:
```python
weight_quantizer = getattr(self.module, 'weight_quantizer', None)
if weight_quantizer is None:
    return False
extra_args = getattr(weight_quantizer, 'backend_extra_args', None)
```

File: modelopt/torch/quantization/model_calib.py [hunk 3]
```diff
@@ -1665,6 +1680,10 @@ def gptq(
     print_rank_0("Updating weights using GPTQ algorithm...")
     for handle in gptq_handles.values():
         handle.update_weights(block_size, perc_damp)
+
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/model_calib.py]: Important: This `weight_quantizer.disable()` after GPTQ is a behavioral change that affects **all** GPTQ paths (not just vector LUT). Previously, after GPTQ, the weight quantizer remained enabled, meaning subsequent forward passes would re-quantize the already-QDQ'd weights.

The fix seems correct in principle (weights are already quantized), but it's a significant behavioral change. Is this intentional for all formats (NVFP4, FP8, INT4)? The existing `test_gptq_updates` and `test_gptq_e2e_flow` tests should be verified to still pass with this change. Also, the `hasattr` check is unnecessary since GPTQ only runs on quantized linears which always have `weight_quantizer`.

File: tests/gpu/torch/quantization/test_gptq_vq.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    )
+
+
+# ---------------------------------------------------------------------------
+# Tests
+# ---------------------------------------------------------------------------
+
+SEED = 42
+OUT_FEATURES = 64
+IN_FEATURES = 128
```
Comment (cjluo-nv) [inline tests/gpu/torch/quantization/test_gptq_vq.py]: Tests import `from luts import clip_vector_prescaled, clip_vector_scalesign_fast` which is not a standard Python package and doesn't appear to be defined anywhere in this repository. This means all three tests will fail with `ModuleNotFoundError` in standard CI environments.

Please either:
1. Add a `pytest.importorskip('luts')` at the top of the file
2. Use `@pytest.mark.skipif` decorators on the tests
3. Or document the dependency and how to install it

Without this, the test file is effectively non-functional for most environments.

File: tests/gpu/torch/quantization/test_gptq_vq.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    nmse = compute_output_nmse(W_orig, Q, A)
+    return Q, nmse
+
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+
+
+def _make_codebook_file(tmp_dir, name, n_codewords, vector_size):
```
Comment (cjluo-nv) [inline tests/gpu/torch/quantization/test_gptq_vq.py]: `torch.save({"sorted_values": codebook}, ...)` in `_make_codebook_file` — this is fine for tests, but note that whoever loads this file will need `weights_only=False` in `torch.load()` (since it's a dict containing a tensor), which is flagged by the repo's security guidelines. Ensure the production `load_vector_lut_codebook` function uses `weights_only=True` or `safetensors`.

File: examples/llm_ptq/hf_ptq.py [hunk 6]
```diff
@@ -419,10 +420,15 @@ def load_model(args: argparse.Namespace):
             attn_implementation=args.attn_implementation,
         )
     else:
-        assert args.qformat in QUANT_CFG_CHOICES, (
-            f"Quantization format is not supported for low memory mode. Supported formats: {QUANT_CFG_CHOICES.keys()}"
-        )
-        quant_cfg = QUANT_CFG_CHOICES[args.qformat]
+        if args.qformat in QUANT_CFG_CHOICES:
+            quant_cfg = QUANT_CFG_CHOICES[args.qformat]
```
Comment (cjluo-nv) [inline examples/llm_ptq/hf_ptq.py]: The `getattr(mtq, args.qformat)` fallback could return any attribute of the `mtq` module (functions, classes, constants) if the format name happens to match. Consider validating that the result is actually a dict (quant config):

```python
elif hasattr(mtq, args.qformat):
    quant_cfg = getattr(mtq, args.qformat)
    if not isinstance(quant_cfg, dict):
        raise ValueError(
            f"mtq.{args.qformat} is not a valid quantization config (got {type(quant_cfg).__name__})"
        )
```

This same pattern is duplicated at line ~1078. Consider extracting to a helper function to avoid duplication.

File: examples/llm_ptq/example_utils.py [hunk 7]
```diff
@@ -583,7 +583,7 @@ def get_model(
     model_kwargs = config_kwargs.copy()
     # Don't set torch_dtype for VILA models as they handle it explicitly in their builder
     if "vila" not in ckpt_path.lower():
-        model_kwargs.setdefault("dtype", "auto")
+        model_kwargs.setdefault("torch_dtype", "auto")
```
Comment (cjluo-nv) [inline examples/llm_ptq/example_utils.py]: Good fix: `dtype` → `torch_dtype` matches the HuggingFace `from_pretrained` API.

However, note there's still `dtype="auto"` on line ~641 inside the `has_pack_quantized_config` block:
```python
model = AutoModelForCausalLM.from_pretrained(
    ckpt_path,
    device_map="auto",
    trust_remote_code=trust_remote_code,
    dtype="auto",  # <-- should this also be torch_dtype?
)
```
Is this intentional?

File: modelopt/torch/quantization/utils/calib_utils.py [hunk 8]
```diff
@@ -224,6 +244,58 @@ def _blockwise_update(self, block_size):
...
+        for blk_start in range(0, num_cols, block_size):
+            blk_end = min(blk_start + block_size, num_cols)
+            errs = torch.zeros_like(w[:, blk_start:blk_end])
+
+            for j in range(blk_start, blk_end, vector_size):
+                d = min(vector_size, blk_end - j)
+                s = quantizer._psx_scales[:, j // quant_block_size].contiguous()
+
+                sub = w[:, j : j + d].contiguous()
+                if d < vector_size:
```
Comment (cjluo-nv) [inline modelopt/torch/quantization/utils/calib_utils.py]: In the inner loop, `w[:, col:blk_end].addr_(err, h_inv[col, col:blk_end], alpha=-1)` updates the working copy `w` in-place. When `d < vector_size` (padding case), the quantizer sees a padded sub-vector and returns a full-sized `q_sub`. The loop only uses `q_sub[:, :d]`, which is correct. However, the padded zeros could affect the quantizer's codeword selection for the last valid elements in the padded vector, potentially introducing different quantization compared to a non-padded path. Is this edge case tested? (It would occur when `blk_end - j < vector_size`.)

**Review by realAsma — APPROVED** [review]

**Review by cjluo-nv — APPROVED** [review]
This is a small, focused PR (+39 -4, 3 files) that adds two clean changes:

1. **KV cache reset in sequential calibration** (`model_calib.py`): Prevents KV cache from leaking across forward replays during GPTQ's multi-pass calibration. The implementation correctly handles both `DynamicCache` (`.reset()`) and older cache formats (set to `None`), and properly shallow-copies `kwargs_input` to avoid mutating captured references.

2. **Backend-specific GPTQ helper registry** (`calib_utils.py` + `model_calib.py`): Adds `_GPTQ_HELPER_REGISTRY` and `register_gptq_helper()` to allow backend-specific GPTQ subclasses to be registered and dispatched. The lookup in `_make_gptq_handle` correctly falls back to the default `GPTQHelper` when no backend is set or no custom helper is registered. The `backend` attribute access via `getattr(m.weight_quantizer, "backend", None)` is correct per the `TensorQuantizer` implementation.

Most previous review comments were from an earlier, larger iteration of the PR that included vector LUT implementation, test files, and example changes. Those files have been removed from this diff, making the previous critical comments no longer applicable. The current diff is infrastructure-only, setting up the extension point for backend-specific GPTQ helpers (presumably to be used by the vector LUT GPTQ in a follow-up).
